### PR TITLE
Ensure promtail config is reset everytime the observability bundle resource version is updated

### PR DESCRIPTION
**Pull request by QuentinBisson**
Towards https://github.com/giantswarm/default-apps-aws/pull/378

This PR use controller runtime watches to be able to rerun a reconciliation loop if the observability-bundle app upgrades from a release prior to 1.0.0 to a release greater than 1.0.0 to fix a race condition issues related to the breaking changes in the bundle.

### CODE REVIEW

File: internal/controller/predicates/predicates.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	if err != nil {
+		return false
+	}
+	newAppVersion, err := semver.Parse(newApp.Spec.Version)
+	if err != nil {
+		return false
+	}
+	breakingVersion := semver.MustParse("1.0.0")
+
+	fmt.Printf("%v\n", oldApp)
```
Comment (QuentinBisson) [inline internal/controller/predicates/predicates.go:49]: This was tested on a CAPI WC Upgrade:

```
&{{ } {quentin-observability-bundle  org-giantswarm  f2e879d8-aafe-428c-b3c7-6357d0afb397 845757342 2 2024-01-10 11:23:35 +0000 UTC <nil> <nil> map[app-operator.giantswarm.io/version:0.0.0 app.giantswarm.io/branch: app.giantswarm.io/commit: app.kubernetes.io/instance:quentin-default-apps app.kubernetes.io/managed-by:Helm app.kubernetes.io/name:default-apps-aws app.kubernetes.io/version:0.40.0 application.giantswarm.io/team:phoenix giantswarm.io/cluster:quentin giantswarm.io/managed-by:quentin-default-apps giantswarm.io/organization:giantswarm giantswarm.io/service-type:managed helm.sh/chart:default-apps-aws-0.40.0] map[app-operator.giantswarm.io/latest-configmap-version:845757330 meta.helm.sh/release-name:quentin-default-apps meta.helm.sh/release-namespace:org-giantswarm] [] [operatorkit.giantswarm.io/app-operator-app] [{chart-operator Update application.giantswarm.io/v1alpha1 2024-01-10 11:23:34 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{".":{},"f:meta.helm.sh/release-name":{},"f:meta.helm.sh/release-namespace":{}},"f:labels":{".":{},"f:app-operator.giantswarm.io/version":{},"f:app.giantswarm.io/branch":{},"f:app.giantswarm.io/commit":{},"f:app.kubernetes.io/instance":{},"f:app.kubernetes.io/managed-by":{},"f:app.kubernetes.io/name":{},"f:app.kubernetes.io/version":{},"f:application.giantswarm.io/team":{},"f:giantswarm.io/cluster":{},"f:giantswarm.io/managed-by":{},"f:giantswarm.io/organization":{},"f:giantswarm.io/service-type":{},"f:helm.sh/chart":{}}},"f:spec":{".":{},"f:catalog":{},"f:config":{".":{},"f:configMap":{".":{},"f:name":{},"f:namespace":{}}},"f:kubeConfig":{".":{},"f:inCluster":{}},"f:name":{},"f:namespace":{},"f:version":{}}} } {logging-operator Update application.giantswarm.io/v1alpha1 2024-01-10 11:23:36 +0000 UTC FieldsV1 {"f:spec":{"f:config":{"f:secret":{".":{},"f:name":{},"f:namespace":{}}},"f:extraConfigs":{},"f:install":{},"f:kubeConfig":{"f:context":{".":{},"f:name":{}},"f:secret":{".":{},"f:name":{},"f:namespace":{}}},"f:namespaceConfig":{},"f:rollback":{},"f:uninstall":{},"f:upgrade":{},"f:userConfig":{".":{},"f:configMap":{".":{},"f:name":{},"f:namespace":{}},"f:secret":{".":{},"f:name":{},"f:namespace":{}}}}} } {app-operator Update application.giantswarm.io/v1alpha1 2024-01-10 11:23:38 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{"f:app-operator.giantswarm.io/latest-configmap-version":{}},"f:finalizers":{".":{},"v:\"operatorkit.giantswarm.io/app-operator-app\"":{}}}} } {app-operator Update application.giantswarm.io/v1alpha1 2024-01-10 11:23:39 +0000 UTC FieldsV1 {"f:status":{".":{},"f:appVersion":{},"f:release":{".":{},"f:lastDeployed":{},"f:status":{}},"f:version":{}}} status}]} {default  {{quentin-cluster-values org-giantswarm} { }} [{configMap quentin-observability-bundle-logging-extraconfig org-giantswarm 25}] {false nil} {true {} { }} observability-bundle org-giantswarm {map[] map[]} {nil} {nil} {nil} {{ } { }} 0.10.1} { {2024-01-10 11:23:40 +0000 UTC  deployed} 0.10.1}}
&{{ } {quentin-observability-bundle  org-giantswarm  f2e879d8-aafe-428c-b3c7-6357d0afb397 845757753 3 2024-01-10 11:23:35 +0000 UTC <nil> <nil> map[app-operator.giantswarm.io/version:0.0.0 app.giantswarm.io/branch: app.giantswarm.io/commit: app.kubernetes.io/instance:quentin-default-apps app.kubernetes.io/managed-by:Helm app.kubernetes.io/name:default-apps-aws app.kubernetes.io/version:0.40.0-3dae9ad41a998f6468a4d45198c5b121ee4f96da application.giantswarm.io/team:phoenix giantswarm.io/cluster:quentin giantswarm.io/managed-by:quentin-default-apps giantswarm.io/organization:giantswarm giantswarm.io/service-type:managed helm.sh/chart:default-apps-aws-0.40.0-3dae9ad41a998f6468a4d45198c5b121ee4f96d] map[app-operator.giantswarm.io/latest-configmap-version:845757330 meta.helm.sh/release-name:quentin-default-apps meta.helm.sh/release-namespace:org-giantswarm] [] [operatorkit.giantswarm.io/app-operator-app] [{chart-operator Update application.giantswarm.io/v1alpha1 2024-01-10 11:23:34 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{".":{},"f:meta.helm.sh/release-name":{},"f:meta.helm.sh/release-namespace":{}},"f:labels":{".":{},"f:app-operator.giantswarm.io/version":{},"f:app.giantswarm.io/branch":{},"f:app.giantswarm.io/commit":{},"f:app.kubernetes.io/instance":{},"f:app.kubernetes.io/managed-by":{},"f:app.kubernetes.io/name":{},"f:app.kubernetes.io/version":{},"f:application.giantswarm.io/team":{},"f:giantswarm.io/cluster":{},"f:giantswarm.io/managed-by":{},"f:giantswarm.io/organization":{},"f:giantswarm.io/service-type":{},"f:helm.sh/chart":{}}},"f:spec":{".":{},"f:catalog":{},"f:config":{".":{},"f:configMap":{".":{},"f:name":{},"f:namespace":{}}},"f:kubeConfig":{".":{},"f:inCluster":{}},"f:name":{},"f:namespace":{},"f:version":{}}} } {logging-operator Update application.giantswarm.io/v1alpha1 2024-01-10 11:23:36 +0000 UTC FieldsV1 {"f:spec":{"f:config":{"f:secret":{".":{},"f:name":{},"f:namespace":{}}},"f:extraConfigs":{},"f:install":{},"f:kubeConfig":{"f:context":{".":{},"f:name":{}},"f:secret":{".":{},"f:name":{},"f:namespace":{}}},"f:namespaceConfig":{},"f:rollback":{},"f:uninstall":{},"f:upgrade":{},"f:userConfig":{".":{},"f:configMap":{".":{},"f:name":{},"f:namespace":{}},"f:secret":{".":{},"f:name":{},"f:namespace":{}}}}} } {app-operator Update application.giantswarm.io/v1alpha1 2024-01-10 11:23:38 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{"f:app-operator.giantswarm.io/latest-configmap-version":{}},"f:finalizers":{".":{},"v:\"operatorkit.giantswarm.io/app-operator-app\"":{}}}} } {app-operator Update application.giantswarm.io/v1alpha1 2024-01-10 11:23:39 +0000 UTC FieldsV1 {"f:status":{".":{},"f:appVersion":{},"f:release":{".":{},"f:lastDeployed":{},"f:status":{}},"f:version":{}}} status}]} {default  {{quentin-cluster-values org-giantswarm} { }} [{configMap quentin-observability-bundle-logging-extraconfig org-giantswarm 25}] {false nil} {true {} { }} observability-bundle org-giantswarm {map[] map[]} {nil} {nil} {nil} {{ } { }} 1.0.0} { {2024-01-10 11:23:40 +0000 UTC  deployed} 0.10.1}}
true

```

File: internal/controller/capicluster_controller.go [hunk 2]
```diff
@@ -81,5 +86,18 @@ func (r *CapiClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 func (r *CapiClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&capi.Cluster{}).
+		// This ensures we run the reconcile loop when the observability-bundle app resource version changes.
+		Watches(
+			&appv1alpha1.App{},
+			handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, object client.Object) []reconcile.Request {
+				return []reconcile.Request{
+					{NamespacedName: types.NamespacedName{
+						Name:      object.GetLabels()["giantswarm.io/cluster"],
```
Comment (QuentinBisson) [inline internal/controller/capicluster_controller.go:95]: This gets the cluster CR for CAPI clusters

File: internal/controller/predicates/predicates.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// Update implements default UpdateEvent filter for validating resource version change.
+func (ObservabilityBundleAppVersionChangedPredicate) Update(e event.UpdateEvent) bool {
+	if e.ObjectOld == nil {
+		return false
+	}
+	if e.ObjectNew == nil {
+		return false
+	}
+
+	if !strings.Contains(e.ObjectOld.GetName(), "observability-bundle") || !strings.Contains(e.ObjectNew.GetName(), "observability-bundle") {
```
Comment (QuentinBisson) [inline internal/controller/predicates/predicates.go:26]: We only watch observability bundles

File: internal/controller/predicates/predicates.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	var oldApp, newApp *appv1alpha1.App
+	var ok bool
+	if oldApp, ok = e.ObjectOld.(*appv1alpha1.App); !ok {
+		return false
+	}
+	if newApp, ok = e.ObjectNew.(*appv1alpha1.App); !ok {
+		return false
+	}
+
+	oldAppVersion, err := semver.New(oldApp.Spec.Version)
```
Comment (QuentinBisson) [inline internal/controller/predicates/predicates.go:39]: We check the app versions

File: internal/controller/vintagemc_controller.go [hunk 5]
```diff
@@ -94,5 +99,18 @@ func (r *VintageMCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 func (r *VintageMCReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&corev1.Service{}).
+		// This ensures we run the reconcile loop when the observability-bundle app resource version changes.
+		Watches(
+			&appv1alpha1.App{},
+			handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, object client.Object) []reconcile.Request {
+				return []reconcile.Request{
+					{NamespacedName: types.NamespacedName{
```
Comment (QuentinBisson) [inline internal/controller/vintagemc_controller.go:107]: This configures the kubernetes service

File: internal/controller/vintagewc_controller.go [hunk 6]
```diff
@@ -81,5 +87,18 @@ func (r *VintageWCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 func (r *VintageWCReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&capiv1beta1.Cluster{}).
+		// This ensures we run the reconcile loop when the observability-bundle app resource version changes.
+		Watches(
+			&appv1alpha1.App{},
+			handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, object client.Object) []reconcile.Request {
+				return []reconcile.Request{
+					{NamespacedName: types.NamespacedName{
+						Name:      object.GetLabels()["giantswarm.io/cluster"],
```
Comment (QuentinBisson) [inline internal/controller/vintagewc_controller.go:96]: This creates a reconcile requests for a vintage WC (needs to be tested once we can have a 1.0.0 release)

File: internal/controller/predicates/predicates.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	if err != nil {
+		return false
+	}
+	newAppVersion, err := semver.Parse(newApp.Spec.Version)
```
Comment (marieroque) [inline internal/controller/predicates/predicates.go:43]: why are you using `semver.New` for oldAppVersion and `semver.Parse` for newAppVersion ?

Comment (QuentinBisson) [reply internal/controller/predicates/predicates.go:43]: Well, because I forgot to change

Comment (QuentinBisson) [reply internal/controller/predicates/predicates.go]: Fixed

**Comment by QuentinBisson** [comment]
I'll test it again, this time on CAPI but I'd love if you could take a look as well :)

**Comment by marieroque** [comment]
> I'll test it again, this time on CAPI but I'd love if you could take a look as well :)

Yes I'll do it 
👁️

**Comment by QuentinBisson** [comment]
Thank you :)

**Comment by QuentinBisson** [comment]
Test successfull on vintage WC but I'll let you test again Marie :)

**Pull request merged**